plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php

```diff
@@ -0,0 +1,283 @@
+			$calling_file = debug_backtrace()[1]['file'] ?? null; // [1], not [0], because the immediate caller is the install_plugin method.
+			if ( ! StringUtil::starts_with( $calling_file, WC_ABSPATH . 'includes/' ) && ! StringUtil::starts_with( $calling_file, WC_ABSPATH . 'src/' ) ) {
+				throw new \InvalidArgumentException( "If the value of 'installed_by' is 'WooCommerce', the caller of the method must be a WooCommerce core class or function." );
+			}
+		}
+
+		if ( ! class_exists( \Automatic_Upgrader_Skin::class ) ) {
+			include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader-skin.php';
+			include_once ABSPATH . 'wp-admin/includes/class-automatic-upgrader-skin.php';
+		}
+		$skin = new \Automatic_Upgrader_Skin();
+
+		if ( ! class_exists( \Plugin_Upgrader::class ) ) {
+			include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+		}
+		$upgrader = new \Plugin_Upgrader( $skin );
+
+		$install_ok = $upgrader->install( $plugin_url );
+
+		$result = array( 'messages' => $skin->get_upgrade_messages() );
+
+		if ( $install_ok ) {
+			$plugin_name    = $upgrader->plugin_info();
+			$plugin_version = get_plugins()[ $plugin_name ]['Version'];
+
+			$result['plugin_name'] = $plugin_name;
+			$plugin_data           = array(
+				'version' => $plugin_version,
+				'date'    => current_time( 'mysql' ),
+			);
+			if ( ! empty( $metadata ) ) {
+				$plugin_data['metadata'] = $metadata;
+			}
+
+			$post_install = function () use ( $plugin_name, $plugin_version, $installed_by, $plugin_url, $plugin_data ) {
+				$auto_installed_plugins                 = get_option( 'woocommerce_autoinstalled_plugins', array() );
+				$auto_installed_plugins[ $plugin_name ] = $plugin_data;
+				$log_context                            = array(
+					'source'        => 'plugin_auto_installs',
+					'recorded_data' => $plugin_data,
+				);
+				update_option( 'woocommerce_autoinstalled_plugins', $auto_installed_plugins );
+				wc_get_logger()->info( "Plugin $plugin_name v{$plugin_version} installed by $installed_by, source: $plugin_url", $log_context );
+			};
+		} else {
+			$messages     = $skin->get_upgrade_messages();
+			$post_install = function () use ( $plugin_url, $installed_by, $messages ) {
+				$log_context = array(
+					'source'             => 'plugin_auto_installs',
+					'installer_messages' => $messages,
+				);
+				wc_get_logger()->error( "$installed_by failed to install plugin from source: $plugin_url", $log_context );
+			};
+		}
+
+		$this->run_callback_in_all_sites( $post_install );
+
+		$result['install_ok'] = $install_ok ?? false;
+		return $result;
+	}
+
+	/**
+	 * Run a callback in each existing site (if multisite) or just once (if single site).
+	 *
+	 * @param callable $callback The callback to run.
+	 */
+	private static function run_callback_in_all_sites( callable $callback ) {
+		if ( ! is_multisite() ) {
+			$callback();
+			return;
+		}
+
+		foreach ( get_sites() as $site ) {
+			switch_to_blog( $site->blog_id );
+			try {
```

Comment:
Should we check that WC is active on those sites before attempting to run an arbitrary callback? I believe this is not a problem with the _current_ callbacks as they will just silently fail at using the WC logger, thanks being wrapped in `try {}`, but this might not necessarily be ok depending on what other possible callbacks do exactly.

I don't feel this is a blocker at this point, but maybe something to consider. What are your thoughts @Konamiman?

Reply:
> Should we check that WC is active on those sites before attempting to run an arbitrary callback?

That's a good point, I have pushed a commit doing that.